spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/dsl/GraphGeneratorVisitor.java

```diff
@@ -0,0 +1,651 @@
+					looseEnd.label);
+			// They should have the same 'flow' as the sequence they are injected into
+			tt.flow = transitionTargets.get(0).flow;
+			tt.lastNodeId = transitionTargets.get(0).lastNodeId;//nodeIds.get(looseEnd.lastNodeId);
+			if (DEBUG) {
+				System.out.println("Remapped "+looseEnd+" to "+tt);
+			}
+			rewrittenTransitions.add(tt);
+		}
+		mainSequence.outstandingTransitions.addAll(rewrittenTransitions);
+		// The copy of this secondary sequence is being inserted into a particular flow.
+		Flow flowBeingInsertedInto = transitionTargets.get(0).flow;
+		Map<String,Node> relevantFlowMapToUpdate = mainSequence.labeledNodesInEachFlow.get(flowBeingInsertedInto);
+		Map<Flow,Map<String,Node>> labeledNodesInSequenceBeingInlined = sequence.labeledNodesInEachFlow;
+		Flow primaryFlowInSequenceBeingInlined = sequence.primaryFlow;
+		for (Map.Entry<Flow, Map<String,Node>> entry: labeledNodesInSequenceBeingInlined.entrySet()) {
+			if (entry.getKey() == primaryFlowInSequenceBeingInlined) {
+				// these should be remapped
+				for (Map.Entry<String,Node> entry2: entry.getValue().entrySet()) {
+					Node n2 = new Node(nodeIds.get(entry2.getKey()),entry2.getValue().name);
+					n2.setLabel(entry2.getValue().getLabel());
+					relevantFlowMapToUpdate.put(entry2.getKey(), n2);
+				}
+			}
+			else {
+				// these can just be copied direct
+				Map<String,Node> newMap = new HashMap<>();
+				for (Map.Entry<String,Node> entry2: entry.getValue().entrySet()) {
+					Node n2 = new Node(nodeIds.get(entry2.getKey()),entry2.getValue().name);
+					n2.setLabel(entry2.getValue().getLabel());
+					newMap.put(entry2.getKey(), n2);
+				}
+				mainSequence.labeledNodesInEachFlow.put(entry.getKey(), newMap);
+			}
+		}
+		
+	}
+
+	@Override
+	public boolean preVisit(Split split) {
+		List<String> open = currentContext().getDanglingNodes();
+		String startId = (open.size() == 0 ? currentContext().startNodeId : open.get(0));
+		// If there are multiple open nodes, we need a sync node !
+		if (open.size() > 1) {
+			String syncId = nextId();
+			Node node = new Node(syncId, "SYNC");
+			addNode(node);
+			for (String openid : open) {
+				addLink(new Link(openid, syncId));
+			}
+			startId = syncId;
+		}
+		contexts.push(new Context(false, true, startId, split));
+		return true;
+	}
+
+	@Override
+	public void postVisit(Split split) {
+		List<String> openAtEndOfFlow = currentContext().getDanglingNodes();
+		contexts.pop();
+		currentContext().addDanglingNodes(true, openAtEndOfFlow);
+	}
+
+	@Override
+	public boolean preVisit(Flow flow) {
+		contexts.push(new Context(true, false, currentContext().startNodeId, flow));
+		currentSequence().primaryFlow = flow;
+		return true;
+	}
+
+	@Override
+	public void postVisit(Flow flow) {
+		// What label references were not resolved within the flow?
+		List<Context.TransitionTarget> transitionTargets = currentContext().getTransitionTargets();
+		for (Context.TransitionTarget tt : transitionTargets) {
```

Comment:
its magic.
Comments added. I do worry about the complexity of graphgeneratorvisitor, I suspect you hit similar complexity in your visitor that builds the real task domain objects.